// routes/Calendar/Calendar.tsx
import * as React from 'react';
import { Dispatch, SetStateAction, useEffect, useState } from 'react';
import {
  AppointmentModel,
  ChangeSet,
  EditingState,
  IntegratedEditing,
  SchedulerDateTime,
  ViewState,
} from '@devexpress/dx-react-scheduler';
import Container from '@material-ui/core/Container';
import {
  AllDayPanel,
  AppointmentForm,
  Appointments,
  AppointmentTooltip,
  ConfirmationDialog,
  CurrentTimeIndicator,
  DateNavigator,
  DayView,
  MonthView,
  Scheduler,
  TodayButton,
  Toolbar,
  ViewSwitcher,
  WeekView,
} from '@devexpress/dx-react-scheduler-material-ui';
import { appointments } from 'routes/Calendar/data';

export type CalState = {
  count: number;
  data: AppointmentModel[];
  currentDate: SchedulerDateTime;
};

export type CallProps = {};
// export type CallAction = {
//   type: 'increment' | 'decrement';
//   payload?: SchedulerDateTime | AppointmentModel[];
// };

//
// function reducer(state: CalState, action: CallAction) {
//   switch (action.type) {
//     case 'increment':
//       return { count: state.count + 1 };
//     case 'decrement':
//       return { count: state.count - 1 };
//     default:
//       throw new Error();
//   }
// }

function commitChanges(
  { added, changed, deleted }: ChangeSet,
  data: AppointmentModel[],
  cb: Dispatch<SetStateAction<AppointmentModel[]>>,
): void {
  console.dir({ added, changed, deleted });
  if (added) {
    const { id }: AppointmentModel = data[data.length - 1];

    const startingAddedId = id !== undefined ? parseInt(`${id}`, 10) + 1 : 0;
    // const startingAddedId = data.length > 0 ? data[data.length - 1].id + 1 : 0;
    cb([...data, { id: startingAddedId, ...(added as AppointmentModel) }]);
  }
  if (changed) {
    cb(
      data.map((appointment) => {
        const { id }: AppointmentModel = appointment;
        const changedItem = id !== undefined ? changed[parseInt(`${id}`, 10)] : undefined;
        return changedItem !== undefined ? { ...appointment, ...changedItem } : appointment;
      }),
    );
  }
  if (deleted !== undefined) {
    cb(data.filter((appointment) => appointment.id !== deleted));
  }
}

export const Calendar: React.FunctionComponent = (): JSX.Element => {
  const [data, setData] = useState<AppointmentModel[]>([]);
  const [currentDate, setCurrentDate] = useState<SchedulerDateTime>(new Date());

  useEffect(() => {
    Promise.resolve(appointments).then((appointments) =>
      setData(
        appointments.map((app, index) => ({
          id: index,
          ...app,
        })),
      ),
    );

    // return (): void => { console.log('useEffect stop'); };
  }, []);

  const currentDateChange = (currentDate: Date): void => {
    setCurrentDate(currentDate);
  };

  return (
    <Container maxWidth="lg">
      <Scheduler data={data} height={660}>
        <ViewState currentDate={currentDate} onCurrentDateChange={currentDateChange} defaultCurrentViewName="Week" />
        <WeekView startDayHour={9} endDayHour={19} />
        <WeekView name="work-week" displayName="Work Week" excludedDays={[0, 6]} startDayHour={9} endDayHour={19} />
        <MonthView />
        <DayView startDayHour={9} endDayHour={19} />
        <Toolbar />
        <DateNavigator />
        <ViewSwitcher />
        <TodayButton />
        <Appointments />
        <AllDayPanel />
        <CurrentTimeIndicator shadePreviousCells={false} shadePreviousAppointments={true} updateInterval={60000} />
        <EditingState onCommitChanges={(changes: ChangeSet): void => commitChanges(changes, data, setData)} />
        <IntegratedEditing />
        <AppointmentTooltip showCloseButton showOpenButton showDeleteButton />
        <AppointmentForm />

        <ConfirmationDialog />
      </Scheduler>
    </Container>
  );
};

export default Calendar;
